service/matching/workers/registry_impl.go

```diff
@@ -0,0 +1,278 @@
+	m := newRegistryImpl(
+		defaultBuckets,
+		defaultEntryTTL,
+		defaultMinEvictAge,
+		defaultMaxEntries,
+		defaultEvictionInterval,
+	)
+
+	lc.Append(fx.Hook{
+		OnStart: func(ctx context.Context) error {
+			m.Start()
+			return nil
+		},
+		OnStop: func(ctx context.Context) error {
+			m.Stop()
+			return nil
+		},
+	})
+
+	return m
+}
+
+func newRegistryImpl(numBuckets int,
+	ttl time.Duration,
+	minEvictAge time.Duration,
+	maxItems int64,
+	evictionInterval time.Duration,
+) *registryImpl {
+	m := &registryImpl{
+		buckets:          make([]*bucket, numBuckets),
+		maxItems:         maxItems,
+		tTL:              ttl,
+		minEvictAge:      minEvictAge,
+		evictionInterval: evictionInterval,
+		hasher:           &maphash.Hash{},
+		quit:             make(chan struct{}),
+	}
+	m.hasher.Reset()
+
+	for i := range m.buckets {
+		m.buckets[i] = newBucket()
+	}
+	return m
+}
+
+// bucketFor hashes the namespace to select a bucket.
+func (m *registryImpl) getBucket(nsID namespace.ID) *bucket {
+	m.hasher.Reset()
+	m.hasher.WriteString(nsID.String()) //nolint:revive
+	hs := m.hasher.Sum64()
+	idx := int(hs % uint64(len(m.buckets)))
+
+	return m.buckets[idx]
+}
+
+// upsertHeartbeat records or refreshes a WorkerHeartbeat under the given namespace.
+// New entries increment the global counter.
+func (m *registryImpl) upsertHeartbeat(nsID namespace.ID, hb *workerpb.WorkerHeartbeat) {
+	b := m.getBucket(nsID)
+	if newEntry := b.upsertHeartbeat(nsID, hb); newEntry {
+		m.total.Add(1)
+	}
+}
+
+// listWorkers returns all WorkerHeartbeats in a namespace
+// for which predicate(hb) returns true.
+func (m *registryImpl) filterWorkers(
+	nsID namespace.ID,
+	predicate func(*workerpb.WorkerHeartbeat) bool,
+) []*workerpb.WorkerHeartbeat {
+	b := m.getBucket(nsID)
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	mp := b.namespaces[nsID]
+	if mp == nil {
+		return nil
+	}
+	out := make([]*workerpb.WorkerHeartbeat, 0, len(mp))
+	for _, e := range mp {
+		if predicate(e.hb) {
+			out = append(out, e.hb)
+		}
+	}
+	return out
+}
+
+// evictLoop periodically triggers TTL and capacity-based eviction.
+func (m *registryImpl) evictLoop() {
+	ticker := time.NewTicker(m.evictionInterval)
+	for {
+		select {
+		case <-ticker.C:
+			m.evictByTTL()
+			m.evictByCapacity()
+		case <-m.quit:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+// evictByTTL removes expired entries across all buckets.
+func (m *registryImpl) evictByTTL() {
+	expireBefore := time.Now().Add(-m.tTL)
+	var removed int64
+	for _, b := range m.buckets {
+		removed += int64(b.evictByTTL(expireBefore))
+	}
+	if removed > 0 {
+		m.total.Add(-removed)
+	}
+}
+
+// evictByCapacity removes entries older than MinEvictAge until under capacity.
+func (m *registryImpl) evictByCapacity() {
+	for m.total.Load() > m.maxItems {
+		removedAny := false
+		threshold := time.Now().Add(-m.minEvictAge)
+		for _, b := range m.buckets {
+			if m.total.Load() <= m.maxItems {
+				return
+			}
+			b.mu.Lock()
+			front := b.order.Front()
```

Comment:
there's no fairness across namespaces here? I guess that's fine for this purpose